src/Symfony/Component/DependencyInjection/Compiler/ServiceLocatorTagPass.php

```diff
@@ -48,7 +50,51 @@ protected function processValue($value, $isRoot = false)
             }
             $arguments[0][$k] = new ServiceClosureArgument($v);
         }
+        ksort($arguments[0]);
 
-        return $value->setArguments($arguments);
+        $value->setArguments($arguments);
+
+        if ($public = $value->isPublic()) {
+            $value->setPublic(false);
+        }
+        $id = 'service_locator.'.md5(serialize($value));
+
+        if ($isRoot) {
+            if ($id !== $this->currentId) {
+                $this->container->setAlias($id, new Alias($this->currentId, $public));
+            }
+
+            return $value;
+        }
+
+        $this->container->setDefinition($id, $value);
+
+        return new Reference($id);
+    }
+
+    /**
+     * @param Reference[] $refMap
```

Comment:
the phpdoc does not match the signature (which can confuse some IDEs which will ask consumers to pass the refMap as first argument)